### PR TITLE
Perf topic:

### PR DESCRIPTION
**Pull request by gluck**
- bump .Net to 4
- change dict key to a struct to reduce allocation
- remove deprecated cecil calls
- small array helper (performs better than linq)

### CODE REVIEW

File: ILRepack/ILRepack.cs [hunk 1]
```diff
@@ -467,7 +468,7 @@ private object ImportCustomAttributeValue(object obj, IGenericParameterProvider
             if (obj is CustomAttributeArgument)
                 return Copy((CustomAttributeArgument)obj, context);
             if (obj is CustomAttributeArgument[])
-                return ((CustomAttributeArgument[])obj).Select(a => Copy(a, context)).ToArray();
+                return ((CustomAttributeArgument[])obj).Clone(a => Copy(a, context));
```
Comment (hazzik) [inline ILRepack/ILRepack.cs]: You can use `Array.ConvertAll` here, which does exactly this

Comment (gluck) [inline ILRepack/ILRepack.cs]: Didn't know it, thanks !

File: ILRepack/MappingHandler.cs [hunk 2]
```diff
@@ -27,13 +27,8 @@ public override int GetHashCode()
                 return scope.GetHashCode() + name.GetHashCode();
             }
 
-            public override bool Equals(object obj)
```
Comment (timotei) [inline ILRepack/MappingHandler.cs]: According to IEquatable doc, they say we should also override Equals to get consistent behaviour (https://msdn.microsoft.com/en-us/library/ms131187%28v=vs.110%29.aspx)

Comment (gluck) [inline ILRepack/MappingHandler.cs]: They also say we should override == and != :smile: 
This isn't a general-purpose type (it should be private), hence I'm not sure it's necessary to add those, they'll never be used.

**Comment by tephe** [comment]
:+1:  Ended up into this class myself while running performance tests. 

Another performance bottleneck that I see here is that we create a new object at each GetRemappedType call. Not sure what impact it does have since now we are using a value type. Perhaps a simple indexing by reference.Scope.Name + reference.FullName?
```
        internal TypeDefinition GetRemappedType(TypeReference r)
        {
            TypeDefinition other;
            if (r.Scope != null && mappings.TryGetValue(GetTypeKey(r), out other))
            {
                return other;
            }
            return null;
        }
```

**Comment by gluck** [comment]
Creating a new string key by concat'ing both would be much worse. Using IEquatable struct is allocation-free, much better.
As of the hashCode, usually indeed we avoid that by doing `hash(A)+prime*hash(B)`.
The only impact is more collision chance on hashes if A and B are within the same value-space, not really the case here (Assembly name vs type name).

Thx !

**Pull request merged**